### PR TITLE
Optimize job order execution in circleci config for shorter build times

### PR DESCRIPTION
**Pull request by addb**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -61,36 +72,229 @@ jobs:
...
 
       # Code coverage upload using Codecov
       # See options explanation here: https://docs.codecov.com/docs/codecov-uploader
       - codecov/upload:
-           upload_name: Coverage Report for tvOS
-           xtra_args: -c -v --xc --xp tvosresults.xcresult
+          upload_name: Coverage Report for tvOS Integration Tests
+          xtra_args: -c -v --xc --xp tvosintegrationtestresults.xcresult
 
+  build:
```
Comment (kevinlind) [inline .circleci/config.yml:292]: Name this step something more descriptive, such as `buildXCFramework`.

File: .circleci/config.yml [hunk 2]
```diff
@@ -61,36 +72,229 @@ jobs:
...
+          name: Repo Update & Install CocoaPods
+          command: make ci-pod-install
+
+      # save pods related files
+      - save_cache:
+          name: Saving CocoaPods Cache
+          key: cocoapods-cache-v1-{{ arch }}-{{ .Branch }}-{{ checksum "Podfile.lock" }}
+          paths:
+            - ./Pods
+            - ~/.cocoapods
```
Comment (kevinlind) [inline .circleci/config.yml:117]: It looks like these steps to install CocoaPods and update the Pods are common to all the jobs. Can you look into using reusable configs in CircleCI or YAML anchors and aliases to define these steps once and reuse in each job?
Here are some references:
https://preventdirectaccess.com/reuse-steps-circleci/
CircleCI reusable config guide: https://circleci.com/docs/reusing-config/
YAML anchors and aliases: https://circleci.com/blog/circleci-hacks-reuse-yaml-in-your-circleci-config-with-yaml/

File: Makefile [hunk 3]
```diff
@@ -40,7 +44,7 @@ clean:
 	(rm -rf build)
 
 build-app:
-	#make -C SampleApps/$(APP_NAME) build-shallow
+	#make -C TestApps/$(APP_NAME) build-shallow
```
Comment (emdobrin) [inline Makefile:47]: can this command be enabled now?

File: Makefile [hunk 4]
```diff
@@ -65,6 +79,43 @@ test-tvos:
...
+	@echo "######################################################################"
+	xcodebuild test -workspace $(PROJECT_NAME).xcworkspace -scheme $(INTEGRATION_TEST_SCHEME) -destination 'platform=iOS Simulator,name=iPhone 8' -derivedDataPath build/outn -resultBundlePath iosintegrationtestresults.xcresult -enableCodeCoverage YES
+
+integration-test-tvos:
+	@echo "######################################################################"
+	@echo "### Integration Testing tvOS"
+	@echo "######################################################################"
+	xcodebuild test -workspace $(PROJECT_NAME).xcworkspace -scheme $(INTEGRATION_TEST_SCHEME) -destination 'platform=tvOS Simulator,name=Apple TV' -derivedDataPath build/outn -resultBundlePath tvosintegrationtestresults.xcresult -enableCodeCoverage YES
+
+
```
Comment (kevinlind) [inline Makefile:118]: You are not referencing these individual test tasks in the CI configuration. I'd say to remove these if they are not used in CI as developers can run these directly from Xcode.

**Review by kevinlind — COMMENTED** [review]
The Makefile still references the "tools" directory. Can you remove the `setup-tools` and `install-githook` rules from the Makefile? You can also do this in another PR if you like.

**Comment by kevinlind** [comment]
> The Makefile still references the "tools" directory. Can you remove the `setup-tools` and `install-githook` rules from the Makefile? You can also do this in another PR if you like.

Better still, instead of removing, add a `pre-commit` hook file which runs SwiftLint, add this to a folder named `.githooks`, then update the `install-githook` rule to `git config core.hooksPath .githooks`. You can see the [Edge](https://github.com/adobe/aepsdk-edge-ios/tree/dev/.githooks) project for an example.

**Comment by addb** [comment]
> git config core.hooksPath .githooks

Thanks for the info. Added it!

**Comment by addb** [comment]
> > The Makefile still references the "tools" directory. Can you remove the `setup-tools` and `install-githook` rules from the Makefile? You can also do this in another PR if you like.
> 
> Better still, instead of removing, add a `pre-commit` hook file which runs SwiftLint, add this to a folder named `.githooks`, then update the `install-githook` rule to `git config core.hooksPath .githooks`. You can see the [Edge](https://github.com/adobe/aepsdk-edge-ios/tree/dev/.githooks) project for an example.

> git config core.hooksPath .githooks

Thanks for the info. Added it!

File: .githooks/pre-commit [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+SWIFTLINT=./Pods/SwiftLint/swiftlint
+CONFIG=.swiftlint.yml
+
+if ! command -v "${SWIFTLINT}" &> /dev/null; then
+	echo "${SWIFTLINT} is not installed. Please run 'make install-swiftlint'."
```
Comment (kevinlind) [inline .githooks/pre-commit:6]: nit: This error is no longer accurate. You could change it to "Please run 'pod install'."

**Review by kevinlind — APPROVED** [review]

**Pull request merged**